aip/general/0161.md

```diff
@@ -0,0 +1,131 @@
+---
+id: 161
+state: reviewing
+created: 2021-03-01
+placement:
+  category: design-patterns
+  order: 85
+---
+
+# Field masks
+
+Often, when updating resources (using an update method as defined in AIP-134 or
+something reasonably similar), it is desirable to specify exactly which fields
+are being updated, so that the service can ignore the rest, even if the user
+sends new values.
+
+It is tempting to define a mask format to handle the precise needs for each
+API. However, masking requirements evolve, and therefore it is prudent to use a
+structured syntax. This allows updates to be able to be made transparently,
+without waiting for UI or client updates.
+
+## Guidance
+
+Update methods (AIP-134) **must** include an `update_mask` field, and Get and
+List methods **may** include a `read_mask` field. These are collectively called
+"field masks", and they use the `google.protobuf.FieldMask` type.
+
+Field masks **must** always be relative to the resource:
+
+```proto
+message UpdateBookRequest {
+  // The book to update.
+  //
+  // The book's `name` field is used to identify the book to be updated.
+  // Format: publishers/{publisher}/books/{book}
+  Book book = 1 [(google.api.field_behavior) = REQUIRED];
+
+  // The list of fields to be updated.
+  // Fields are specified relative to the book
+  // (e.g. `title`, `rating`; *not* `book.title` or `book.rating`).
+  google.protobuf.FieldMask update_mask = 2;
+}
+```
+
+### Read-write consistency
+
+Read and write behavior for field masks **must** be self-consistent if a mask
+is present:
+
+- If a user updates a resource with a given mask, and then reads the same
+  resource with the same mask, the exact same data **must** be returned.
+- Similarly, reading a resource with a given mask and then updating the
+  resource with the returned data and the same mask **must** be a no-op.
+
+**Note:** This implies that any mask that is valid for either read or write
+**must** be valid for both.
+
+### Specifying specific fields
+
+Field masks **must** permit the specification of specific fields in a defined
+struct, using the `.` character for traversal.
+
+Because field masks are always relative to the resource, direct fields on the
+resource require no traversal (examples: `title`, `rating`). Traversal is used
+when resources contain messages (example: `author.given_name`).
+
+**Note:** A user **must** be able to specify either a field as a whole, or one
+of its subfiends: `author` and `author.given_name` are both valid.
+
+### Map fields
+
+Field masks **may** permit the specification of specific fields in a map, if
+and only if the map's fields are either strings or integers, using the `.`
+character for traversal.
```

Comment:
It's been a long time since I've looked at the internal documents - I'd expected it to be `[]`. Was that *only* for repeated fields? (Explicitly prohibited here...)

Reply:
It is `.` everywhere as best as I can tell.